benchmarks/src/main/java/io/grpc/benchmarks/qps/AsyncServer.java

```diff
@@ -184,36 +191,194 @@ public ForkJoinWorkerThread newThread(ForkJoinPool pool) {
   }
 
   public static class BenchmarkServiceImpl extends BenchmarkServiceGrpc.BenchmarkServiceImplBase {
+    private static final long SHUTDOWN_POLL_NANOS = TimeUnit.MILLISECONDS.toNanos(100);
+    // Always use the same canned response for bidi. This is allowed by the spec.
+    private static final int BIDI_RESPONSE_BYTES = 100;
+    private static final Messages.SimpleResponse BIDI_RESPONSE = Messages.SimpleResponse
+        .newBuilder()
+        .setPayload(Messages.Payload.newBuilder()
+            .setBody(ByteString.copyFrom(new byte[BIDI_RESPONSE_BYTES])).build())
+        .build();
+
+    private final ThreadLocal<Long> shutdownPollNanos = new ThreadLocal<Long>();
+    private final AtomicBoolean shutdown = new AtomicBoolean();
+
+    public BenchmarkServiceImpl() {
+    }
+
+    public void shutdown() {
+      this.shutdown.set(true);
+    }
+
+    /**
+     * Helper method to avoid polling the shutdown volatile in a tight loop. The server shutdown
+     * operation only happens once at the end of the test scenario, so there's no need to
+     * constantly check it.
+     */
+    private boolean shouldShutdown() {
+      Long latestPoll = shutdownPollNanos.get();
```

Comment:
+1, this seems unnecessary. Also, `System.nanoTime()` involves some `volatile` operations itself and has other overheads.